daemon/logger/loggertest/logreader.go

```diff
@@ -381,6 +383,75 @@ func (tr Reader) TestFollow(t *testing.T) {
 		<-doneReading
 		assert.DeepEqual(t, logs, expected, compareLog)
 	})
+
+	t.Run("Concurrent", tr.TestConcurrent)
+}
+
+// TestConcurrent tests the Logger and its LogReader implementation for
+// race conditions when logging from multiple goroutines concurrently.
+func (tr Reader) TestConcurrent(t *testing.T) {
+	t.Parallel()
+	l := tr.Factory(t, logger.Info{
+		ContainerID:   "logconcurrent0",
+		ContainerName: "logconcurrent123",
+	})(t)
+
+	msgLine := func(source string, i int) []byte {
+		return []byte(fmt.Sprintf("%s msg %d", source, i))
+	}
+	time0, _ := time.Parse("2006-01-02", "2000-01-01")
+	msgTime := func(i int) time.Time {
+		return time0.Add(time.Hour * time.Duration(i))
+	}
+
+	wg := &sync.WaitGroup{}
+	logTo := func(source string, howMany int) {
+		defer wg.Done()
+		for i := 0; i < howMany; i += 1 {
+			msg := logger.NewMessage()
+			msg.Line = msgLine(source, i)
+			msg.Source = source
+			msg.Timestamp = msgTime(i)
+			l.Log(msg)
+		}
+	}
+
+	// Read all logs
+	lw := l.(logger.LogReader).ReadLogs(logger.ReadConfig{Follow: true, Tail: -1})
+	defer lw.ConsumerGone()
+
+	const messagesCount = 10000
+	// Log concurrently from two sources
+	wg.Add(2)
+	go logTo("stdout", messagesCount)
+	go logTo("stderr", messagesCount)
+	go func() {
+		defer l.Close()
+		wg.Wait()
+	}()
+
+	lastStdout := 0
+	lastStderr := 0
+
+	// Check if the message count, order and content is equal to what was logged
+	for l := range lw.Msg {
```

Comment:
Use the `readMessage` helper rather than looping directly through the `lw.Msg` channel as the helper handles timeouts and checking the `lw.Err` channel. (The `LogWatcher` API leaves much to be desired and could really use a refactor.) I suspect that the `msg="error closing current log file" error="invalid argument"` message you saw was accompanied by an error in the `lw.Err` channel but we don't know what the error was since the channel is never read from.
```suggestion
	for {
		l := readMessage(t, lw)
		if l == nil {
			break
		}